go/gen.py

```diff
@@ -13,20 +23,34 @@
 go_extractor_dir = workspace_dir / "go" / "extractor"
 go_dbscheme = workspace_dir / "go" / "ql" / "lib" / "go.dbscheme"
 r = runfiles.Create()
-go, gazelle, go_gen_dbscheme = map(r.Rlocation, sys.argv[1:])
+go, gazelle, go_gen_dbscheme = map(r.Rlocation, opts.generators)
+
 
-print("updating vendor")
+if opts.force:
+    print("clearing vendor directory")
+    shutil.rmtree(go_extractor_dir / "vendor")
+
+existing_build_files = set(go_extractor_dir.glob("*/**/BUILD.bazel"))
+
+print("updating vendor directory")
 subprocess.check_call([go, "-C", go_extractor_dir, "work", "vendor"])
 
-print("clearing generated BUILD files")
-for build_file in go_extractor_dir.glob("*/**/BUILD.bazel"):
-    build_file.unlink()
+if opts.force:
+    print("clearing generated BUILD files")
+    for build_file in existing_build_files:
+        build_file.unlink()
 
 print("running gazelle")
 subprocess.check_call([gazelle])
 
+build_files_to_update = set(go_extractor_dir.glob("*/**/BUILD.bazel"))
+if not opts.force:
+    build_files_to_update -= existing_build_files
+    # these are always refreshed
+    build_files_to_update.update(go_extractor_dir.glob("vendor/**/BUILD.bazel"))
+
```

Comment:
Minor: This part may benefit from a comment explaining the intended logic. I had to think a bit about the code in the entire file to understand what's going on here.